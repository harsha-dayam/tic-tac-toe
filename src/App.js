import React from 'react';
import './App.scss';

class App extends React.Component {
  constructor(props) {
	super(props);
	this.state = {
		// 1 stands for X and 0 stands for O
		currentValue: 1,
		box: [
			['', '', ''],
			['', '', ''],
			['', '', '']
		]
	}
	this._handleClick = this.handleClick.bind(this); 
	this._setUserValue = this.setUserValue.bind(this);  
	this._checkWinner = this.checkWinner.bind(this);
  }

  checkWinner() {
    // console.log(this.state.box);
  }

  setUserValue (el, value) {
	const { box } = this.state,
		  boxCopy = JSON.parse(JSON.stringify(box)),
		  x = el.dataset.row,
		  y = el.dataset.column; // deep cloning since its an array or arrays
		  
	el.classList.add("value" + value);
	boxCopy[x][y] = value;

	this.setState({
	  	box: boxCopy
	}, ()=> {
		this.checkWinner();
	})
  }

  handleClick(e) {
	const { currentValue } = this.state,
		el = e.target;
	debugger;
			
	if (!el.classList.contains("valueX") && !el.classList.contains("valueO")) {
		if (currentValue === 1) {
			this.setUserValue(el, "X");
		} else {
			this.setUserValue(el, "O");
		}

		this.setState({
			currentValue: currentValue ^ 1
		})
	}
  }

  render() {
		return (
			<div className="tic-tac-toe">
				<table id="board">
					<tbody>
						{[[1, 2, 3].map((value, index1) => (
							// 3 rows
							<tr key={index1}>
								{[1, 2, 3].map((value, index2) => 
									// 3 cells in each row
									<td key={index2} data-row={index1} data-column={index2} onClick={this._handleClick}></td>
								)}
							</tr>
						))]}
					</tbody>
				</table>
			</div>
		);
  	}
}  

export default App;
