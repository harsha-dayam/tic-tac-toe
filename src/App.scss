.tic-tac-toe {
  text-align: center;

  table {
    width: 50%;
    border-collapse: collapse;
    text-align: center;
    margin: 0 auto;

    td {
      border: 6px solid #222;
      height: 150px;
      width: 150px;
  
      &::after {
        content: "";
        font-size: 30px;
      }
  
      &:first-of-type {
        border-left-color: transparent;
        border-top-color: transparent;
      }
  
      &:nth-of-type(2) {
        border-top-color: transparent;
      }
  
      &:nth-of-type(3) {
        border-right-color: transparent;
        border-top-color: transparent;
      }
    }

    tr:nth-of-type(3) td {
      border-bottom-color: transparent;
    }


    .value1::after {
      content: "X";
      font-size: 30px;
    }

    .value0::after {
      content: "O";
      font-size: 30px;
    }
  }  
}