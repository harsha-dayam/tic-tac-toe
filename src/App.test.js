import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow , mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

it('renders without crashing', () => {
	const div = document.createElement('div');
	ReactDOM.render(<App />, div);
	ReactDOM.unmountComponentAtNode(div);
});

describe('App', () => {
	const wrapper = shallow(
		<App />
	);
	let param = {
		target: {
			classList: {
				contains: jest.fn(),
				add: jest.fn()
			},
			dataset: {
				row: 0,
				column: 0
			}
		}
	};

	it('should render correctly', () => {		
		expect(shallowToJson(wrapper)).toMatchSnapshot();
	});

	it('click on first tile', () => {
		const arr = [
			['', '', ''],
			['', '', ''],
			['', '', '']
		], 
		arr1 = [
			['X', '', ''],
			['', '', ''],
			['', '', '']
		];

		expect(wrapper.state().currentValue).toEqual(1);
		expect(wrapper.state().box).toEqual(arr);
		wrapper.find('td').first().simulate('click', param );	
		expect(wrapper.state().currentValue).toEqual(0);
		expect(wrapper.state().box).toEqual(arr1);
	});

	it('click on second tile upon clicking first tile', () => {
		const arr = [
			['X', '', ''],
			['', '', ''],
			['', '', '']
		], 
		arr1 = [
			['X', 'O', ''],
			['', '', ''],
			['', '', '']
		];

		param.target.dataset.column = 1; // second tile (0, 1)

		expect(wrapper.state().currentValue).toEqual(0);
		expect(wrapper.state().box).toEqual(arr);
		wrapper.find('td').at(2).simulate('click', param );	
		expect(wrapper.state().currentValue).toEqual(1);
		expect(wrapper.state().box).toEqual(arr1);
	});
	
	it('click on first tile again', () => {
		const arr = [
			['X', 'O', ''],
			['', '', ''],
			['', '', '']
		], 
		arr1 = [
			['X', 'O', ''],
			['', '', ''],
			['', '', '']
		];

		// element already has a class assigned i.e a value assigned to it
		param.target.classList.contains = jest.fn().mockReturnValueOnce(true);

		expect(wrapper.state().currentValue).toEqual(1);
		expect(wrapper.state().box).toEqual(arr);
		wrapper.find('td').first().simulate('click', param);	
		expect(wrapper.state().currentValue).toEqual(1);
		expect(wrapper.state().box).toEqual(arr1);
	});
});

