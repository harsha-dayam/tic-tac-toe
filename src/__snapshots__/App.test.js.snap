// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div
  className="tic-tac-toe"
>
  <table
    id="board"
  >
    <tbody>
      <tr
        key="0"
      >
        <td
          data-column={0}
          data-row={0}
          key="0"
          onClick={[Function]}
        />
        <td
          data-column={1}
          data-row={0}
          key="1"
          onClick={[Function]}
        />
        <td
          data-column={2}
          data-row={0}
          key="2"
          onClick={[Function]}
        />
      </tr>
      <tr
        key="1"
      >
        <td
          data-column={0}
          data-row={1}
          key="0"
          onClick={[Function]}
        />
        <td
          data-column={1}
          data-row={1}
          key="1"
          onClick={[Function]}
        />
        <td
          data-column={2}
          data-row={1}
          key="2"
          onClick={[Function]}
        />
      </tr>
      <tr
        key="2"
      >
        <td
          data-column={0}
          data-row={2}
          key="0"
          onClick={[Function]}
        />
        <td
          data-column={1}
          data-row={2}
          key="1"
          onClick={[Function]}
        />
        <td
          data-column={2}
          data-row={2}
          key="2"
          onClick={[Function]}
        />
      </tr>
    </tbody>
  </table>
</div>
`;
